bootstore/src/schemes/v0/peer.rs

```diff
@@ -0,0 +1,802 @@
+
+    // Clone for use by connection tasks to send to the main peer task
+    conn_tx: mpsc::Sender<ConnToMainMsg>,
+}
+
+impl From<Config> for FsmConfig {
+    fn from(value: Config) -> Self {
+        FsmConfig {
+            learn_timeout: value.learn_timeout,
+            rack_init_timeout: value.rack_init_timeout,
+            rack_secret_request_timeout: value.rack_secret_request_timeout,
+        }
+    }
+}
+
+impl Peer {
+    pub fn new(config: Config, log: &Logger) -> (Peer, PeerHandle) {
+        // We only expect one outstanding request at a time for `Init_` or
+        // `LoadRackSecret` requests, We can have one of those requests in
+        // flight while allowing `PeerAddresses` updates.
+        let (tx, rx) = mpsc::channel(3);
+
+        // There are up to 31 sleds sending messages. These are mostly one at a
+        // time for each sled, but we leave some extra room.
+        let (conn_tx, conn_rx) = mpsc::channel(128);
+        let fsm =
+            Fsm::new_uninitialized(config.id.clone(), config.clone().into());
+        let id_str = config.id.to_string();
+        (
+            Peer {
+                config,
+                fsm,
+                peers: BTreeSet::new(),
+                handle_unique_id_counter: 0,
+                shutdown: false,
+                accepted_connections: BTreeMap::new(),
+                negotiating_connections: BTreeMap::new(),
+                connections: BTreeMap::new(),
+                rx,
+                init_responder: None,
+                rack_secret_responder: None,
+                log: log
+                    .new(o!("component" => "bootstore", "peer_id" => id_str)),
+                conn_rx,
+                conn_tx,
+            },
+            PeerHandle { tx },
+        )
+    }
+
+    /// Run the main loop of the peer
+    ///
+    /// This should be spawned into its own tokio task
+    pub async fn run(&mut self) {
+        // select among timer tick/received messages
+        let mut interval = interval(self.config.time_per_tick);
+        interval.set_missed_tick_behavior(MissedTickBehavior::Delay);
+        let listener = TcpListener::bind(&self.config.addr).await.unwrap();
+        while !self.shutdown {
+            tokio::select! {
+                res = listener.accept() => self.on_accept(res).await,
+                Some(request) = self.rx.recv() => {
+                    self.on_api_request(request).await;
+                }
+                Some(msg) = self.conn_rx.recv() => self.on_conn_msg(msg).await,
+                _ = interval.tick() => {
+                    if let Err(errors) = self.fsm.tick(Instant::now().into()) {
+                        for (_, err) in errors {
+                            self.handle_api_error(err).await;
+                        }
+                    }
+                    // Even with errors there may be messages that need sending
+                    self.deliver_envelopes().await;
```

Comment:
Would it make sense to move this into the main body of the loop?  i.e.
```rust
while !self.shutdown {
    tokio::select! {
        // ...
    }
    self.deliver_envelopes().await
}
```
Then, we don't need to remember it in every branch of `on_conn_msg` (and it's presumably safe / cheap to call if no envelopes are queued up)

Reply:
Good idea. Fixed in 851cc853f